File: src/main/java/org/ffpy/plugin/coding/action/menu/NewEnumCodeAction.java
package org.ffpy.plugin.coding.action.menu;

import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.LangDataKeys;
import com.intellij.openapi.editor.CaretModel;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.Editor;
import com.intellij.psi.PsiDirectory;
import com.intellij.psi.PsiFile;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.ffpy.plugin.coding.action.BaseAction;
import org.ffpy.plugin.coding.constant.TemplateName;
import org.ffpy.plugin.coding.util.*;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.function.Predicate;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

/**
 * 生成EnumCode
 */
@Slf4j
public class NewEnumCodeAction extends BaseAction {

    private static final Predicate<String> PREDICATE_UPDATE = Pattern.compile(
            "\\s*`\\w+`\\s+(TINY)?INT\\(\\d*\\)\\s+.*COMMENT\\s+'.+[:：].*'.*",
            Pattern.MULTILINE | Pattern.CASE_INSENSITIVE).asPredicate();

    private static final Pattern PATTERN_CREATE_TABLE = Pattern.compile(
            "CREATE\\s+TABLE\\s+`(\\w+)`", Pattern.CASE_INSENSITIVE);

    private static final Predicate<String> PREDICATE_CREATE_TABLE = Pattern.compile(
            ".*CREATE\\s+TABLE.*", Pattern.CASE_INSENSITIVE).asPredicate();

    private static final Pattern PATTERN_COLUMN_NAME = Pattern.compile(
            "^\\s*`(\\w+)`");

    private static final Pattern PATTERN_COMMENT = Pattern.compile(
            "COMMENT\\s+'.*[:：](.+)'", Pattern.CASE_INSENSITIVE);

    @Override
    public void action() throws Exception {
        PsiDirectory directory = getDirectory()
                .orElseThrow(() -> new RuntimeException("找不到constant文件夹"));

        PsiFile file = createFile(directory);
        if (file == null) return;

        env.getWriteActions().add(() -> {
            FileUtils.addIfAbsent(directory, file);
            FileUtils.navigateFile(env.getProject(), directory, file.getName());
        }).run();
    }

    @Override
    public void update(@NotNull AnActionEvent e) {
        ActionShowHelper.of(e)
                .fileNameMatch(".*\\.sql")
                .and(() -> PREDICATE_UPDATE.test(getLineText(e)))
                .update();
    }

    /**
     * 获取字段名
     *
     * @param lineText 当前行
     */
    private String getColumnName(String lineText) {
        Matcher matcher = PATTERN_COLUMN_NAME.matcher(lineText);
        return matcher.find() ? matcher.group(1) : "";
    }

    /**
     * 获取注释
     */
    private String getComment(String lineText) {
        Matcher matcher = PATTERN_COMMENT.matcher(lineText);
        return matcher.find() ? matcher.group(1) : "";
    }

    /**
     * 获取类名
     *
     * @param lineText 当前行
     */
    private String getClassName(String lineText) {
        return MyStringUtils.toTitle(getTableName().toLowerCase()) +
                MyStringUtils.toTitle(getColumnName(lineText).toLowerCase());
    }

    /**
     * 获取模板参数
     *
     * @param lineText  当前行
     * @param className 类名
     */
    private Map<String, Object> getParams(String lineText, String className) {
        Map<String, Object> params = new HashMap<>();
        params.put("className", className);
        params.put("items", getItems(getComment(lineText)));
        return params;
    }

    /**
     * 获取项目列表
     */
    private List<Item> getItems(String comment) {
        List<Item> items = Arrays.stream(StringUtils.split(comment, ",，"))
                .map(item -> {
                    String[] split = StringUtils.split(item, "-");
                    return split.length == 2 ? new Item(split[0], split[1], split[1]) : null;
                })
                .filter(Objects::nonNull)
                .collect(Collectors.toList());

        // 翻译
        boolean needTranslate = false;
        // 如果没有中文则不需要翻译
        for (Item item : items) {
            if (MyStringUtils.hasChinese(item.getName())) {
                needTranslate = true;
                break;
            }
        }

        if (needTranslate) {
            try {
                String text = items.stream().map(Item::getName).reduce((s1, s2) -> s1 + "。" + s2)
                        .orElse("");
                String result = env.getTranslateHelper().zh2En(text);
                String[] names = StringUtils.split(result, '.');
                for (int i = 0; i < names.length && i < items.size(); i++) {
                    String name = names[i];
                    if (StringUtils.isNotEmpty(name)) {
                        items.get(i).setName(name.trim().toUpperCase().replaceAll("[^\\w]+", "_"));
                    }
                }
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }

        return items;
    }

    /**
     * 获取表名
     */
    private String getTableName() {
        Editor editor = env.getEditor().orElse(null);
        if (editor == null) return "";

        CaretModel caretModel = editor.getCaretModel();
        int line = caretModel.getLogicalPosition().line;
        Document document = editor.getDocument();
        while (line > 0) {
            String text = EditorUtils.getLineText(document, --line);
            if (text == null) {
                return "";
            }
            if (PREDICATE_CREATE_TABLE.test(text)) {
                Matcher matcher = PATTERN_CREATE_TABLE.matcher(text);
                return matcher.find() ? matcher.group(1) : "";
            }
        }
        return "";
    }

    /**
     * 生成Java文件
     *
     * @param directory 文件夹
     */
    private PsiFile createFile(PsiDirectory directory) {
        String lineText = getLineText(env.getEvent());

        String className = getClassName(lineText);
        String filename = className;
        String path = filename + ".java";
        if (directory.findFile(path) != null) {
            NotificationHelper.info("{}已存在", filename).show();
            FileUtils.navigateFile(env.getProject(), directory, path);
            return null;
        }
        return env.createJavaFile(TemplateName.ENUM_CODE, filename,
                getParams(lineText, className));
    }

    /**
     * 获取文件夹
     */
    private Optional<PsiDirectory> getDirectory() {
        return env.getPackageFile()
                .map(file -> file.findFileByRelativePath("/constant"))
                .map(it -> env.getDirectoryFactory().createDirectory(it));
    }

    /**
     * 获取当前行内容
     */
    private String getLineText(@NotNull AnActionEvent e) {
        return Optional.ofNullable(e.getData(LangDataKeys.EDITOR))
                .map(EditorUtils::getCurLineText)
                .orElse("");
    }

    @Data
    @AllArgsConstructor
    public static class Item {

        /** 码值 */
        private String code;

        /** 字段名 */
        private String name;

        /** 注释 */
        private String comment;
    }
}

File: src/main/kotlin/org/ffpy/plugin/coding/action/menu/NewEnumCodeAction.kt
package org.ffpy.plugin.coding.action.menu

import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.LangDataKeys
import com.intellij.openapi.application.WriteAction
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.util.ThrowableComputable
import com.intellij.psi.PsiDirectory
import com.intellij.psi.PsiFile
import org.apache.commons.lang3.StringUtils
import org.ffpy.plugin.coding.action.BaseAction
import org.ffpy.plugin.coding.constant.Constant
import org.ffpy.plugin.coding.constant.TemplateName
import org.ffpy.plugin.coding.ui.dialog.InputDialog
import org.ffpy.plugin.coding.util.*
import org.ffpy.plugin.coding.util.FileUtils.addIfAbsent
import java.io.IOException
import java.util.*
import java.util.regex.Pattern

/**
 * 生成EnumCode
 */
class NewEnumCodeAction : BaseAction() {

    companion object {
        private val PREDICATE_UPDATE = Pattern.compile(
            "\\s*`\\w+`\\s+(TINY)?INT\\(\\d*\\)\\s+.*COMMENT\\s+'.+[:：].*'.*",
            Pattern.MULTILINE or Pattern.CASE_INSENSITIVE
        ).asPredicate()

        private val PATTERN_CREATE_TABLE = Pattern.compile(
            "CREATE\\s+TABLE\\s+`(\\w+)`", Pattern.CASE_INSENSITIVE
        )

        private val PREDICATE_CREATE_TABLE = Pattern.compile(
            ".*CREATE\\s+TABLE.*", Pattern.CASE_INSENSITIVE
        ).asPredicate()

        private val PATTERN_COLUMN_NAME = Pattern.compile(
            "^\\s*`(\\w+)`"
        )

        private val PATTERN_COMMENT = Pattern.compile(
            "COMMENT\\s+'.*[:：](.+)'", Pattern.CASE_INSENSITIVE
        )
    }

    override fun action() {
        InputDialog(
            "生成EnumCode",
            "包名",
            null,
            Constant.PATTERN_PACKAGE_NAME,
            InputDialog.Action { packageName: String? ->
                val directory = getDirectory(packageName!!) ?: return@Action false
                val file = createFile(directory) ?: return@Action true
                env.writeActions.add {
                    addIfAbsent(directory, file)
                    FileUtils.navigateFile(env.project, directory, file.name)
                }.run()
                true
            }).show()
    }

    override fun update(e: AnActionEvent) {
        ActionShowHelper.of(e)
            .fileNameMatch(".*\\.sql")
            .and { PREDICATE_UPDATE.test(getLineText(e)) }
            .update()
    }

    /**
     * 获取字段名
     *
     * @param lineText 当前行
     */
    private fun getColumnName(lineText: String): String {
        val matcher = PATTERN_COLUMN_NAME.matcher(lineText)
        return if (matcher.find()) matcher.group(1) else ""
    }

    /**
     * 获取注释
     */
    private fun getComment(lineText: String): String {
        val matcher = PATTERN_COMMENT.matcher(lineText)
        return if (matcher.find()) matcher.group(1) else ""
    }

    /**
     * 获取类名
     *
     * @param lineText 当前行
     */
    private fun getClassName(lineText: String): String {
        return MyStringUtils.toTitle(getTableName().toLowerCase()) +
                MyStringUtils.toTitle(getColumnName(lineText).toLowerCase())
    }

    /**
     * 获取模板参数
     *
     * @param lineText  当前行
     * @param className 类名
     */
    private fun getParams(lineText: String, className: String): Map<String, Any> {
        val params: MutableMap<String, Any> = HashMap()
        params["className"] = className
        params["items"] = getItems(getComment(lineText))
        return params
    }

    /**
     * 获取项目列表
     */
    private fun getItems(comment: String): List<Item> {
        val items: List<Item> = StringUtils.split(comment, ",，").asSequence()
            .mapNotNull {
                val split = StringUtils.split(it, "-")
                if (split.size == 2) Item(split[0], split[1], split[1]) else null
            }.toList()

        // 翻译
        var needTranslate = false
        // 如果没有中文则不需要翻译
        for (item in items) {
            if (MyStringUtils.hasChinese(item.name)) {
                needTranslate = true
                break
            }
        }
        if (needTranslate) {
            try {
                val text = items.asSequence().map { it.name }.reduce { acc, s -> "$acc。$s" } ?: ""
                val result: String = TranslateHelper(env.project).zh2En(text)
                val names = StringUtils.split(result, '.')
                var i = 0
                while (i < names.size && i < items.size) {
                    val name = names[i]
                    if (StringUtils.isNotEmpty(name)) {
                        items[i].name = name.trim { it <= ' ' }.toUpperCase().replace("[^\\w]+".toRegex(), "_")
                    }
                    i++
                }
            } catch (e: Exception) {
                throw RuntimeException(e)
            }
        }
        return items
    }

    /**
     * 获取表名
     */
    private fun getTableName(): String {
        val editor: Editor = env.editor ?: return ""
        val caretModel = editor.caretModel
        var line = caretModel.logicalPosition.line
        val document = editor.document
        while (line > 0) {
            val text = EditorUtils.getLineText(document, --line) ?: return ""
            if (PREDICATE_CREATE_TABLE.test(text)) {
                val matcher = PATTERN_CREATE_TABLE.matcher(text)
                return if (matcher.find()) matcher.group(1) else ""
            }
        }
        return ""
    }

    /**
     * 生成Java文件
     *
     * @param directory 文件夹
     */
    private fun createFile(directory: PsiDirectory): PsiFile? {
        val lineText = getLineText(env.event)
        val className = getClassName(lineText)
        val path = "$className.java"
        if (directory.findFile(path) != null) {
            NotificationHelper.info("{}已存在", className).show()
            FileUtils.navigateFile(env.project, directory, path)
            return null
        }
        return env.createJavaFile(
            TemplateName.ENUM_CODE, className,
            getParams(lineText, className)
        )
    }

    /**
     * 获取文件夹
     *
     * @param packageName 包名
     */
    private fun getDirectory(packageName: String): PsiDirectory? {
        return try {
            WriteAction.computeAndWait(ThrowableComputable {
                env.findOrCreateDirectoryByPackageName(
                    packageName
                )
            })
        } catch (e: IOException) {
            throw RuntimeException("创建文件夹失败: " + e.message)
        }
    }

    /**
     * 获取当前行内容
     */
    private fun getLineText(e: AnActionEvent): String {
        return e.getData(LangDataKeys.EDITOR)?.let { EditorUtils.getCurLineText(it) } ?: ""
    }

    data class Item(
        /** 码值  */
        var code: String? = null,

        /** 字段名  */
        var name: String? = null,

        /** 注释  */
        var comment: String? = null
    )
}